File: 2.1_sort3/sort3.cpp
/*
ID: dermotl1
LANG: C++
PROG: sort3
*/

#include <algorithm>
#include <fstream>
#include <vector>

using namespace std;

int main() {
  ofstream fout("sort3.out");
  ifstream fin("sort3.in");

  int n;
  fin >> n;

  vector<int> original(n);
  vector<int> sorted(n);

  for (int i = 0; i < n; i++) {
    fin >> original[i];
    sorted[i] = original[i];
  }

  sort(sorted.begin(), sorted.end());

  int num = 0;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (i == j) continue;
      if (original[i] != sorted[i] && original[j] != sorted[j] &&
          original[i] == sorted[j] && original[j] == sorted[i]) {
        swap(original[i], original[j]);
        num++;
        break;
      }
    }
  }

  int unsort = 0;
  for (int i = 0; i < n; i++) {
    if (original[i] != sorted[i]) unsort++;
  }

  num = (unsort / 3) * 2 + num;  // pairs

  fout << num << "\n";
  return 0;
}

File: 2.1_sort3/sort3.java

/*
ID: dermotl1
LANG: JAVA
PROG: sort3
*/

import java.io.*;
import java.lang.*;
import java.util.*;

public class sort3 {
  public static void main(String[] args) throws FileNotFoundException {
    Scanner in = new Scanner(new File("sort3.in"));
    PrintWriter out = new PrintWriter(new File("sort3.out"));

    int n;
    n = in.nextInt();

    int[] original = new int[n];
    int[] sorted = new int[n];

    for (int i = 0; i < n; i++) {
      original[i] = in.nextInt();
      sorted[i] = original[i];
    }

    Arrays.sort(sorted);

    int num = 0; 
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {

        if (i == j) continue;

        if (original[i] != sorted[i] && original[j] != sorted[j] &&
          original[i] == sorted[j] && original[j] == sorted[i]) {
          
            int temp = original[i];
            original[i] = original[j];
            original[j] = temp;

            num++;
            break;
       }
      }
    }

    int unsort = 0;

    for (int i = 0; i < n; i++) {
      if (original[i] != sorted[i]) unsort++;
   }

    num = (unsort / 3) * 2 + num;

   out.println(num);

   in.close();
   out.close();
  }
}